File: ocaml/11-20/13.ml
(* Run-length encoding of a list (direct solution). (medium)

Implement the so-called run-length encoding data compression method directly. I.e. don't explicitly create the sublists containing the duplicates, as in problem "Pack consecutive duplicates of list elements into sublists", but only count them. As in problem "Modified run-length encoding", simplify the result list by replacing the singleton lists (1 X) by X.
*)


type 'a rle =
  | One of 'a
  | Many of int * 'a


let collect (c, n, zs) x =
  match zs with
  | [] -> (c, n, zs)
  | _  -> if   x = c 
          then (c, (n+1), (Many (n+1, c))::(List.tl zs))
          else (x, 1, (One x)::zs) 

(* encode : 'a list -> list rle *)
let encode = function
  | [] -> []
  | xs -> let x = List.hd xs in
          let (c, n, rs) = List.fold_left collect (x, 0, [One x]) xs in
          List.rev rs

let _ = assert (
  (encode ["a";"a";"a";"a";"b";"c";"c";"a";"a";"d";"e";"e";"e";"e"]) = 
  [Many (4, "a"); One "b"; Many (2, "c"); Many (2, "a"); One "d"; Many (4, "e")])

File: prolog/11-20/13.pl
/*
Implement the so-called run-length encoding data compression method directly. I.e. don't explicitly create the sublists containing the duplicates, as in problem 1.09, but only count them. As in problem 1.11, simplify the result list by replacing the singleton terms [1,X] by X.

Example:
?- encode_direct([a,a,a,a,b,c,c,a,a,d,e,e,e,e],X).
X = [[4,a],b,[2,c],[2,a],d,[4,e]]
*/

singles([], Acc, Acc).
singles([[1, X] | T], Acc, Zs) :- singles(T, [X|Acc], Zs).
singles([[N, X] | T], Acc, Zs) :- 
  N \= 1,
  singles(T, [[N, X]| Acc], Zs).

encd([], Curr, Acc, Encs) :- singles([Curr|Acc], [], Encs).
encd([H|T], [], Acc, Encs) :- encd(T, [1, H], Acc, Encs).
encd([H|T], [N, H], Acc, Encs) :- 
  N1 is N + 1, 
  encd(T, [N1, H], Acc, Encs).
encd([H|T], [N, X], Acc, Encs) :- 
  X \= H,
  encd(T, [1, H], [[N, X] | Acc], Encs).


encode_direct([], []).
encode_direct([H|T], Encs) :- encd([H|T], [], [], Encs).
